File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        from collections import deque
        m = len(grid)
        n = len(grid[0]) 
        queue = deque()
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:
                    queue.append([i, j])
        if not queue:
            for i in range(m):
                for j in range(n):
                    if grid[i][j] == 1:
                        return -1
            return 0
        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]
        curr_time = 3
        while queue:
            x, y = queue.popleft()
            for i in range(4):
                nx = x + dx[i]           
                ny = y + dy[i]
                if 0 <= nx < m and 0 <= ny < n:
                    if grid[nx][ny] == 1 or grid[nx][ny] > grid[x][y]:
                        grid[nx][ny] = grid[x][y] + 1
                        queue.append([nx, ny]) 
        maximum = 2
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    return -1
                maximum = max(maximum, grid[i][j])

        return maximum - 2   

File: 0994-rotting-oranges/0994-rotting-oranges.cpp
class Solution {
private:
    vector<int> dx = {-1, 1, 0, 0};
    vector<int> dy = {0, 0, -1, 1};
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size(); 
        queue<pair<int,int>> q;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 2) q.push({i,j});
            }
        }
        int ans = 0;
        while (!q.empty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                pair<int,int> curr = q.front();
                q.pop();
                int x = curr.first;
                int y = curr.second;
                for (int j = 0; j < 4; j++) {
                    int nx = x + dx[j];
                    int ny = y + dy[j];
                    if (0 <= nx && nx < m && 0 <= ny && ny < n && grid[nx][ny] == 1) {
                        grid[nx][ny] = 2;
                        q.push({nx,ny});
                    }
                }
            }
            if (!q.empty()) ans++;
        }
        for(int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) return -1;
            }
        } 
        return ans;
    }
};